api/iterator.go

```diff
@@ -0,0 +1,60 @@
+package api
+
+import (
+	"fmt"
+	dbm "github.com/tendermint/tm-db"
+)
+
+// This stores all Iterators for one contract
+type Iterators []dbm.Iterator
+
+// IteratorStack contains one entry for each contract, LIFO style
+var IteratorStack []Iterators
+
+// startContract is called at the beginning of a contract runtime to create a new item on the IteratorStack
+func startContract() {
+	IteratorStack = append(IteratorStack, nil)
+	// TODO: remove debug
+	fmt.Printf("startContract: Stack height: %d\n", len(IteratorStack))
+}
+
+// endContract is called at the end of a contract runtime to remove one item from the IteratorStack
+func endContract() {
+	l := len(IteratorStack)
+	var remove Iterators
+	IteratorStack, remove = IteratorStack[:l-1], IteratorStack[l-1]
+	// free all iterators we pop off the stack
+	for _, iter := range remove {
+		fmt.Printf("endContract: close iterator\n")
+		iter.Close()
+	}
```

Comment:
not that familiar with Go `for` loops, but, couldn't you simplify this to not use `range`, and not have the throwaway variable `_`?

Reply:
The way to get items from an array or a map is to use range.

`for k := range foo` will return all keys (or all indexes if an array)

`for k, v := range foo` will return all keys/indexes and values.

It's just how Go is...